routers/api/v1/repo/issue_label.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2016 The Gogs Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package repo
+
+import (
+	api "github.com/gogits/go-gogs-client"
+
+	"github.com/gogits/gogs/models"
+	"github.com/gogits/gogs/modules/context"
+	"github.com/gogits/gogs/routers/api/v1/convert"
+)
+
+func GetIssueLabels(ctx *context.APIContext) {
+	issue, err := models.GetIssueByIndex(ctx.Repo.Repository.ID, ctx.ParamsInt64(":index"))
+	if err != nil {
+		if models.IsErrIssueNotExist(err) {
+			ctx.Status(404)
+		} else {
+			ctx.Error(500, "GetIssueByIndex", err)
+		}
+		return
+	}
+
+	apiLabels := make([]*api.Label, len(issue.Labels))
+	for i := range issue.Labels {
+		apiLabels[i] = convert.ToLabel(issue.Labels[i])
+	}
+
+	ctx.JSON(200, &apiLabels)
+}
+
+func AddIssueLabels(ctx *context.APIContext, form api.IssueLabelsOption) {
+	if !ctx.Repo.IsWriter() {
+		ctx.Status(403)
+		return
+	}
+
+	issue, err := models.GetIssueByIndex(ctx.Repo.Repository.ID, ctx.ParamsInt64(":index"))
+	if err != nil {
+		if models.IsErrIssueNotExist(err) {
+			ctx.Status(404)
+		} else {
+			ctx.Error(500, "GetIssueByIndex", err)
+		}
+		return
+	}
+
+	var labels []*models.Label
+	if labels, err = filterLabelsByRepoID(form.Labels, issue.RepoID); err != nil {
+		ctx.Error(400, "filterLabelsByRepoID", err)
+		return
+	}
+
+	for i := range labels {
+		if !models.HasIssueLabel(issue.ID, labels[i].ID) {
+			if err := models.NewIssueLabel(issue, labels[i]); err != nil {
+				ctx.Error(500, "NewIssueLabel", err)
+				return
+			}
+		}
+	}
+
+	updatedIssue, err := models.GetIssueByIndex(ctx.Repo.Repository.ID, ctx.ParamsInt64(":index"))
+	if err != nil {
+		if models.IsErrIssueNotExist(err) {
+			ctx.Status(404)
+		} else {
+			ctx.Error(500, "GetIssueByIndex", err)
+		}
+		return
+	}
+
+	apiLabels := make([]*api.Label, len(updatedIssue.Labels))
+	for i := range updatedIssue.Labels {
+		apiLabels[i] = convert.ToLabel(updatedIssue.Labels[i])
+	}
+
+	ctx.JSON(200, &apiLabels)
+}
+
+func ReplaceIssueLabels(ctx *context.APIContext, form api.IssueLabelsOption) {
+	if !ctx.Repo.IsWriter() {
+		ctx.Status(403)
+		return
+	}
+
+	issue, err := models.GetIssueByIndex(ctx.Repo.Repository.ID, ctx.ParamsInt64(":index"))
+	if err != nil {
+		if models.IsErrIssueNotExist(err) {
+			ctx.Status(404)
+		} else {
+			ctx.Error(500, "GetIssueByIndex", err)
+		}
+		return
+	}
+
+	var labels []*models.Label
+	if labels, err = filterLabelsByRepoID(form.Labels, issue.RepoID); err != nil {
+		ctx.Error(400, "filterLabelsByRepoID", err)
+		return
+	}
+
+	if err := issue.ClearLabels(); err != nil {
```

Comment:
This removes all existing labels before adding new ones.  Would you rather the replace function check the existing labels and only remove the ones that are not in the new list?